### PR TITLE
Fms2 io update:domain_reads

### LINKED ISSUE #1218: Improve fms2_io performance
**Is your feature request related to a problem? Please describe.**

Doman Reads:
- In the current version of fms2_io, the io_root pe does 1 read for each of the pes and sends it to the other ranks, which is very inefficient

Compressed Writes:
- In the current version of fms2_io, the io_root_pe gathers and writes the data one PE at a time, instead of just doing one write

**Describe the solution you'd like**
Doman Reads:
- The root pe should do 1 read instead of multiple. The sending of the data should be non-blocking.

Compressed Writes:
- The root pe should gather all of the data and do one write instead of multiple.

**Describe alternatives you've considered**
N/A

**Additional context**
N/A

### PR DESCRIPTION
**Pull request by uramirez8707**
**Description**

1. Modified domain_reads_2d and domain_reads_3d:
    - Root pe reads the data
    - Uses mpp_scatter to send the data to the other pes
2. Extended mpp_scatter to work for int8
3.  Added unit tests to test all of the domain_read/domain_write interfaces
4. Domain_reads_4d and domains_reads_5d were unchanged
    - Mpp_scatter only support 2d and 3d arrays
5. Mpp_scatter assumes the data is (x,y,z), which may not always be the case, so I added a kludge for it … 

Fixes #1218 #1219

This should give some performance gain in initialization time for high resolutions.

**How Has This Been Tested?**
CI, SHiELD

The initialization timings for C3072 SHiELD run:
Original: 
Min: 206.756363 Max: 215.601578 Avg: 209.762848

New: 
Min: 153.169189 Max: 161.906982 Avg: 156.292923

**Checklist:**
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] Any dependent changes have been merged and published in downstream modules
- [x] New check tests, if applicable, are included
- [x] `make distcheck` passes

### CODE REVIEW

**Review by ganganoaa — APPROVED** [review]
Looks fine to me.

**Comment by uramirez8707** [comment]
@ganganoaa  thank you for catching those documentation errors.

**Review by bensonr — APPROVED** [review]
The logic looks sound.

**Comment by rem1776** [comment]
@uramirez8707 did we want to add this to main or patch 2023.01 first? there's a`release/2023.01` branch to use if we're patching, its on the same commit as main currently.

**Comment by uramirez8707** [comment]
Let's patch 2023.01 first, so merge this and #1227 to `release/2023.01` and tag as `2023.01.01-alpha1` so we can do more testing.

File: fms2_io/include/domain_read.inc [hunk 1]
```diff
@@ -86,48 +86,52 @@ end subroutine domain_read_1d
...
-  integer(kind=i4_kind), dimension(:,:), allocatable :: buf_i4_kind
-  integer(kind=i8_kind), dimension(:,:), allocatable :: buf_i8_kind
-  real(kind=r4_kind), dimension(:,:), allocatable :: buf_r4_kind
-  real(kind=r8_kind), dimension(:,:), allocatable :: buf_r8_kind
-  logical :: buffer_includes_halos
-  integer :: xgmin !< Starting x index of global io domain
-  integer :: ygmin !< Starting y index of global io domain
+  type(FmsNetcdfDomainFile_t),  intent(in)           :: fileobj         !< File object.
+  character(len=*),             intent(in)           :: variable_name   !< Variable name.
+  class(*), contiguous, target, intent(inout)        :: vdata(:,:)      !< Data that will
```
Comment (bensonr) [inline fms2_io/include/domain_read.inc:91]: I meant to mention your comments mention _writes_ instead of _reads_ here and in the 3d version.

**Review by bensonr — APPROVED** [review]

**Review by rem1776 — APPROVED** [review]

**Pull request merged**